clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -323,11 +323,17 @@ private synchronized void disableHeartbeatThread() {
     }
 
     private void closeHeartbeatThread() {
```

Comment:
Just to clarify, at any given point, `heartbeatThread` may be null, or may not be.  It can also be changed from non-null to null by code running inside the lock, like `startHeartbeatThread`, `disableHeartbeatThread`, `pollHeartbeat`, etc..  The fact that we didn't take the lock before dealing with the variable is the bug being fixed here.

I can move the call to `HeartbeatThread#close` inside the synchronized block.  But the join still needs to be outside, and the join cannot refer to the mutable instance variable, since you need a lock to access that.